annotation/tests/test_annotations.py

```diff
@@ -235,3 +274,156 @@ def test_create_manifest_json(
+        if is_latest:
+            mock_get_pages_sha.assert_called_once_with(
+                doc.pages,
+                doc.base_revision,
+                doc.validated,
+                doc.failed_validation_pages,
+                doc.user,
+            )
+        else:
+            mock_get_pages_sha.assert_called_once_with(
+                doc.pages,
+                latest_doc.revision,
+                doc.validated,
+                doc.failed_validation_pages,
+                doc.user,
+            )
+        mock_check_docs.assert_called_once()
+        if not new_equal_latest:
+            mock_construct_doc_links.assert_called_once()
+            mock_db_add.assert_called_once()
+            mock_db_add_all.assert_called_once_with([])
+            mock_db_commit.assert_called_once()
+            mock_convert_bucket.assert_called_once_with(TEST_TENANT)
+            mock_upload_pages.assert_called_once_with(
+                pages=doc.pages,
+                pages_sha=pages_sha_latest[0],
+                s3_path=s3_path,
+                bucket_name=TEST_TENANT,
+                s3_resource=None,
+            )
+            mock_create_manifest.assert_called_once_with(
+                ANY,
+                s3_path,
+                s3_file_path,
+                s3_file_bucket,
+                TEST_TENANT,
+                annotated_doc.job_id,
+                annotated_doc.file_id,
+                db,
+                None,
+            )
+
+
+def test_construct_annotated_doc_db_error(
+    annotated_doc,
+    annotation_doc_for_save,
+):
+    is_latest = False
+    new_equal_latest = False
+    doc = annotation_doc_for_save
+    latest_doc = annotated_doc
+    s3_file_path = "path"
+    s3_file_bucket = "bucket"
+
+    pages_sha_latest = ({"1": "a"}, "b")
+    pages_sha_non_latest = ({"1": "a"}, "c")
+
+    db = Mock()
+
+    with patch.object(db, "add"), patch.object(db, "add_all"), patch.object(
+        db,
+        "commit",
+        side_effect=IntegrityError(
+            statement="TEST", params=("testuser",), orig=Exception("TESTING")
+        ),
+    ) as mock_db_commit, patch.object(
+        db, "rollback"
+    ) as mock_db_rollback, patch(
+        "annotation.annotations.main.get_pages_sha",
+        return_value=pages_sha_latest if is_latest else pages_sha_non_latest,
+    ), patch(
+        "annotation.annotations.main.check_docs_identity",
+        return_value=new_equal_latest,
+    ), patch(
+        "annotation.annotations.main.construct_document_links", return_value=[]
+    ):
+        with pytest.raises(DuplicateAnnotationError):
+            construct_annotated_doc(
+                db,
+                annotated_doc.user,
+                None,
+                annotated_doc.job_id,
+                annotated_doc.file_id,
+                doc,
+                TEST_TENANT,
+                s3_file_path,
+                s3_file_bucket,
+                latest_doc,
+                1,  # task_id
```

Comment:
Please note that the task_id is known according to the function signature, so the comment explaining it can be omitted.